sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala

```diff
@@ -147,6 +147,9 @@ abstract class Optimizer(catalogManager: CatalogManager)
       RewriteNonCorrelatedExists,
       ComputeCurrentTime,
       GetCurrentDatabaseAndCatalog(catalogManager),
+      // Before replace 'Deduplicate(Keys, Union)' with Aggregate, it needs to
+      // combine all adjacent 'Union' operators into a single 'Union' if necessary.
+      CombineUnions,
       ReplaceDeduplicateWithAggregate) ::
```

Comment:
why not move this rule to the `Replace Operators` batch?

Reply:
I am not sure it's ok to move rule 'ReplaceDeduplicateWithAggregate' to the Replace Operators batch, I will check.

Reply:
Done